src/V3PerPackage/StringLocker.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace NuGet.Services.V3PerPackage
+{
+    public class StringLocker
+    {
+        private readonly object _semaphoresLock = new object();
+        private readonly Dictionary<string, SemaphoreSlim> _semaphores = new Dictionary<string, SemaphoreSlim>();
+
+        private readonly ILogger<StringLocker> _logger;
+
+        public StringLocker(ILogger<StringLocker> logger)
+        {
+            _logger = logger;
+        }
+
+        public async Task<IDisposable> AcquireAsync(string semaphoreName)
+        {
+            SemaphoreSlim semaphore;
+            lock(_semaphores)
+            {
+                if (!_semaphores.TryGetValue(semaphoreName, out semaphore))
+                {
+                    semaphore = new SemaphoreSlim(1);
+                    _semaphores[semaphoreName] = semaphore;
+                }
+            }
+
+            var acquired = await semaphore.WaitAsync(TimeSpan.Zero);
+            if (!acquired)
+            {
+                _logger.LogInformation("The string {SemaphoreName} has already been acquired. Waiting.", semaphoreName);
+                await semaphore.WaitAsync();
+            }
+            
+            return new DisposableSemaphoreSlim(semaphore);
+        }
+
+        public async Task<IDisposable> AcquireAsync(IEnumerable<string> semaphoreNames)
+        {
+            var disposables = new DisposableList();
+            var semaphoreNameList = semaphoreNames.OrderBy(x => x).Distinct().ToList();
+
+            try
+            {
+                foreach (var semaphoreName in semaphoreNameList)
+                {
+                    disposables.Add(await AcquireAsync(semaphoreName));
+                }
+            }
+            catch
+            {
+                disposables.Dispose();
+                throw;
+            }
+
+            return disposables;
+        }
+
+        private class DisposableList : List<IDisposable>, IDisposable
+        {
+            public void Dispose()
+            {
+                foreach (var disposable in this)
+                {
+                    disposable?.Dispose();
+                }
+            }
+        }
+
+        private class DisposableSemaphoreSlim : IDisposable
```

Comment:
>DisposableSemaphoreSlim [](start = 22, length = 23)

What do you think of renaming `_disposed` to `_state` and adding constants for `Live` (0) and `Disposed` (1) states? Also, I would set the initial state to `Disposed` explicitly.

Reply:
Yeah, that sounds fine. `Interlocked` does not work with enums so I have to do some casting, but I think it's still clearer.

Reply:
Fixed.